crates/client/src/registry_url.rs

```diff
@@ -0,0 +1,194 @@
+use anyhow::{anyhow, bail, Context, Result};
+use reqwest::IntoUrl;
+use url::{Host, Url};
+
+/// The base URL of a registry server.
+// Note: The inner Url always has a scheme and host.
+#[derive(Clone)]
+pub struct RegistryUrl(Url);
+
+impl RegistryUrl {
+    /// Parses and validates the given URL into a [`RegistryUrl`].
+    pub fn new(url: impl IntoUrl) -> Result<Self> {
+        // Default to a HTTPS scheme if none is provided
+        let mut url: Url = if !url.as_str().contains("://") {
+            Url::parse(&format!("https://{url}", url = url.as_str()))
+                .context("failed to parse registry server URL")?
+        } else {
+            url.into_url()
+                .context("failed to parse registry server URL")?
+        };
+
+        match url.scheme() {
+            "https" => {}
+            "http" => {
+                // Only allow HTTP connections to loopback
+                match url
+                    .host()
+                    .ok_or_else(|| anyhow!("expected a host for URL `{url}`"))?
+                {
+                    Host::Domain(d) => {
+                        if d != "localhost" {
+                            bail!("an unsecured connection is not permitted to `{d}`");
+                        }
+                    }
+                    Host::Ipv4(ip) => {
+                        if !ip.is_loopback() {
+                            bail!("an unsecured connection is not permitted to address `{ip}`");
+                        }
+                    }
+                    Host::Ipv6(ip) => {
+                        if !ip.is_loopback() {
+                            bail!("an unsecured connection is not permitted to address `{ip}`");
+                        }
+                    }
+                }
+            }
+            _ => bail!("expected a HTTPS scheme for URL `{url}`"),
+        }
+
+        // Normalize by appending a '/' if missing
+        if !url.path().ends_with('/') {
+            url.set_path(&(url.path().to_string() + "/"));
+        }
+
+        Ok(Self(url))
+    }
+
+    /// Returns a mostly-human-readable string that identifies the registry and
+    /// contains only the characters `[0-9a-zA-Z-._]`. This string is
+    /// appropriate to use with external systems that can't accept arbitrary
+    /// URLs such as file system paths.
```

Comment:
See tests for examples.

Reply:
Thanks for catching this issue in the client!